terragraph_planner/optimization/topology_operations.py

```diff
@@ -504,125 +502,104 @@ def compute_capex(topology: Topology, params: OptimizerParams) -> Capex:
     )
 
 
-def _get_max_tx_power_of_all_links(
-    topology: Topology,
-    max_eirp_dbm: Optional[float],
-) -> Dict[str, float]:
-    """
-    Get max tx power for links in the topology.
-    """
-    max_tx_power = {}
-    for link_id, link in topology.links.items():
-        if (
-            link.link_type == LinkType.ETHERNET
-            or link.tx_sector is None
-            or link.rx_sector is None
-        ):
-            max_tx_power[link_id] = -math.inf
-            continue
-
-        tx_data = topology.sites[link.tx_site.site_id].device
-        max_tx_power[link_id] = get_max_tx_power(
-            tx_sector_params=tx_data.sector_params,
-            max_eirp_dbm=max_eirp_dbm,
-        )
-    return max_tx_power
-
-
 def update_link_caps_with_sinr(
-    topology: Topology, maximum_eirp: Optional[float]
+    topology: Topology,
+    params: OptimizerParams,
+    max_iterations: int,
 ) -> None:
     """
     Prior to selecting a subset of active links, we compute link capacities
     and MCS values using the computed SNR values (not SINR). This is because
     the SINR values are directly affected by the subset of active links.
 
     Once the set of active links are selected, we iteratively adjust link
-    budgets to converge. We first compute SINR values for each link using
-    the `analyze_interference` function. Then update the link capacities and
-    MCS values using this newly computed SINR values of each link; next
-    adjust the Tx power based on the new MCS, interference, and Max available
-    Tx power; the link budgets would be re-calculated at last.
-
-    @param topology: A Topology object.
-    @param maximum_eirp: Maximum ERIP.
+    budgets to convergence. In each iteration, we first compute SINR values
+    for each link, update the link capacities and MCS class based on the SINR,
+    adjust the tx power based on the new MCS class and then recompute RSL/SNR
+    based on the updated tx power.
     """
+    planner_assert(
+        max_iterations >= 1,
+        "Number of TPC iterations must be at least 1",
+        OptimizerException,
+    )
 
+    add_link_capacities(topology, params, with_deviation=True)
```

Comment:
This is moved out of post-optimization to reporting so reporting can be called without computing link budgets ahead of time. This also helps ensure such computation is consistent with what's needed within this function.